### PR TITLE
libcurl: add version 8.4.0 (fix CVE-2023-38545)

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **libcurl/***

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by RubenRBS — COMMENTED** [review]
I need to ask for context on the linked issue and ask where things are regarding it before approving, thanks for brining it to my attention :)

**Comment by toge** [comment]
@RubenRBS 
I think following errors are caused by that conan's issue.

```
CMake Error at C:/J2/w/prod-v1/bsr/7946/bbabb/.conan/data/libcurl/8.3.0/_/_/build/fa3bb3edaa54f36a50cf6abdc608eebc812e703c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:22 (add_executable):
  Target "cmTC_01e0c" links to target
  "CONAN_LIB::openssl_OpenSSL_SSL_libssl_RELEASE" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at C:/J2/w/prod-v1/bsr/7946/bbabb/.conan/data/libcurl/8.3.0/_/_/build/fa3bb3edaa54f36a50cf6abdc608eebc812e703c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:22 (add_executable):
  Target "cmTC_01e0c" links to target
  "CONAN_LIB::openssl_OpenSSL_Crypto_libcrypto_RELEASE" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?


CMake Error at C:/J2/w/prod-v1/bsr/7946/bbabb/.conan/data/libcurl/8.3.0/_/_/build/fa3bb3edaa54f36a50cf6abdc608eebc812e703c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:22 (add_executable):
  Target "cmTC_01e0c" links to target "CONAN_LIB::zlib_zlib_RELEASE" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
```
https://c3i.jfrog.io/c3i/misc/logs/pr/19769/1-windows-visual_studio/libcurl/8.3.0//fa3bb3edaa54f36a50cf6abdc608eebc812e703c-build.txt

**Comment by valgur** [comment]
> @RubenRBS 
> I think following errors are caused by that conan's issue.
> 
> ```
> CMake Error at C:/J2/w/prod-v1/bsr/7946/bbabb/.conan/data/libcurl/8.3.0/_/_/build/fa3bb3edaa54f36a50cf6abdc608eebc812e703c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:22 (add_executable):
>   Target "cmTC_01e0c" links to target
>   "CONAN_LIB::openssl_OpenSSL_SSL_libssl_RELEASE" but the target was not
>   found.  Perhaps a find_package() call is missing for an IMPORTED target, or
>   an ALIAS target is missing?
> 
> 
> CMake Error at C:/J2/w/prod-v1/bsr/7946/bbabb/.conan/data/libcurl/8.3.0/_/_/build/fa3bb3edaa54f36a50cf6abdc608eebc812e703c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:22 (add_executable):
>   Target "cmTC_01e0c" links to target
>   "CONAN_LIB::openssl_OpenSSL_Crypto_libcrypto_RELEASE" but the target was
>   not found.  Perhaps a find_package() call is missing for an IMPORTED
>   target, or an ALIAS target is missing?
> 
> 
> CMake Error at C:/J2/w/prod-v1/bsr/7946/bbabb/.conan/data/libcurl/8.3.0/_/_/build/fa3bb3edaa54f36a50cf6abdc608eebc812e703c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:22 (add_executable):
>   Target "cmTC_01e0c" links to target "CONAN_LIB::zlib_zlib_RELEASE" but the
>   target was not found.  Perhaps a find_package() call is missing for an
>   IMPORTED target, or an ALIAS target is missing?
> ```
> https://c3i.jfrog.io/c3i/misc/logs/pr/19769/1-windows-visual_studio/libcurl/8.3.0//fa3bb3edaa54f36a50cf6abdc608eebc812e703c-build.txt

Huh, that's interesting. I'm getting that same error for libjpeg with MSVC in https://github.com/conan-io/conan-center-index/pull/19298. I wonder if it's related in any way.

**Comment by RubenRBS** [comment]
@toge sorry for the late response, didn't see this ping until now

Where does the error come from? A previous build? the current logs seem all to be passing

**Comment by dietssa** [comment]
I noticed that build fails locally for me with the following configuration (buildbot cancelled execution before attempting them):
libcurl/8.4.0@ Windows, Visual Studio, Debug+Release, shared=True

The error message is:
`
libcurl/8.4.0@dietssa/develop (test package): Calling build()
[...]
LINK : fatal error LNK1104: cannot open file 'libcurl_imp.lib' [C:\work\git\libcurl\testbuild\test_package.vcxproj]
`
There is actually no .lib-file in the build folder at all.

In contrast to that, I tested the libcurl upstream cmake build for 8.4.0 as described in https://github.com/curl/curl/blob/master/docs/INSTALL.cmake and it builds the .lib-file as expected.

Diffing the full linker command line in the conan recipe build vs. upstream build showed that the linker option [https://learn.microsoft.com/en-us/cpp/build/reference/def-specify-module-definition-file](/DEF) that is missing in case of conan recipe build. /DEF is used to pass a .def-file with EXPORTS statements, which [triggers the linker to create the import library](https://learn.microsoft.com/en-us/cpp/build/reference/implib-name-import-library?view=msvc-170).

I suspect this is related to the following change in libcurl release 8.4.0: https://github.com/curl/curl/pull/11914

In the conan recipe, we prevent this from working correctly due to a patch that removes the "CurlSymbolHiding"-logic from the upstream CMakeLists.txt:
https://github.com/conan-io/conan-center-index/blob/master/recipes/libcurl/all/conanfile.py#L332-L334

Suggested fix: Remove this patch in conanfile.py. Its marked as "suspicious" anyway ;-)

**Comment by toge** [comment]
@dietssa 
Thank you for your detail investigation!
The `CurlSymbolHiding` is now enabled only for 8.4 to minimize the impact.
If it is not adequate, please tell me again.

**Comment by Cogitri** [comment]
FWIW, I also had to set `HAVE_OPENSSL_SRP` and `HAVE_SSL_CTX_SET_QUIC_METHOD` for my windows build to work.

**Comment by NiuBlibing** [comment]
@toge @RubenRBS Hi, is this pr ready to merge?

**Review by jwillikers — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
Checked the build log https://c3i.jfrog.io/c3i/misc/logs/pr/19769/7-linux-gcc/libcurl/8.4.0//f13ed95cd7c05bdadafca9cb0e33c7e261bae784-build.txt and it looks okay according these new changes.

File: recipes/libcurl/all/conanfile.py [hunk 1]
```diff
@@ -303,7 +303,7 @@ def _patch_autotools(self):
                                   "AC_CHECK_LIB(z,",
                                   f"AC_CHECK_LIB({zlib_name},")
             replace_in_file(self, configure_ac,
-                                  "-lz ",
+                                  "-lz",
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:306]: This change has a good chance to break `with_zstd=True`. There was a good reason for this extra space in pattern.

Comment (johan-boule) [reply recipes/libcurl/all/conanfile.py:306]: Is there an upstream proposal to add the configure script option?

**Pull request merged**